cpp/oneapi/dal/algo/knn/backend/model_impl.hpp

```diff
@@ -18,35 +18,179 @@
 
 #include "oneapi/dal/algo/knn/common.hpp"
 #include "oneapi/dal/algo/knn/backend/model_interop.hpp"
+#include "oneapi/dal/backend/serialization.hpp"
+#include "oneapi/dal/backend/interop/table_conversion.hpp"
+
+#ifdef ONEDAL_DATA_PARALLEL
+#include <src/algorithms/k_nearest_neighbors/oneapi/bf_knn_classification_predict_kernel_ucapi.h>
+#else
+#include <daal/src/algorithms/k_nearest_neighbors/bf_knn_classification_predict_kernel.h>
+#endif
 
 namespace oneapi::dal::knn {
 
+#define KNN_SERIALIZABLE(Task, ClassificationId, SearchId)             \
+    ONEDAL_SERIALIZABLE_MAP2(Task,                                     \
+                             (task::classification, ClassificationId), \
+                             (task::search, SearchId))
+
 template <typename Task>
 class detail::v1::model_impl : public base {
 public:
-    model_impl() : interop_(nullptr) {}
+    model_impl() {}
     model_impl(const model_impl&) = delete;
     model_impl& operator=(const model_impl&) = delete;
 
-    model_impl(backend::model_interop* interop) : interop_(interop) {}
+    virtual backend::model_interop* get_interop() = 0;
+};
+
+namespace detail {
+namespace v1 {
+
+template <typename Task>
+class brute_force_model_impl : public model_impl<Task>,
+                               public KNN_SERIALIZABLE(Task,
+                                                       knn_brute_force_classification_model_impl_id,
+                                                       knn_brute_force_search_model_impl_id) {
+public:
+    brute_force_model_impl() {}
+    brute_force_model_impl(const brute_force_model_impl&) = delete;
+    brute_force_model_impl& operator=(const brute_force_model_impl&) = delete;
+
+    brute_force_model_impl(const table& data, const table& responses)
+            : data(data),
+              responses(responses) {}
+
+    ~brute_force_model_impl() {}
+
+    backend::model_interop* get_interop() override {
+        return nullptr;
+    }
+
+    void serialize(dal::detail::output_archive& ar) const override {
+        ar(data, responses);
+    }
 
-    ~model_impl() {
+    void deserialize(dal::detail::input_archive& ar) override {
+        ar(data, responses);
+    }
+
+    table data;
+    table responses;
```

Comment:
This must be private, to access these variables introduce getter/setter methods

Reply:
Public variables are used in svm and forest model_impl's and it would be inconsistent with these algos using private here. Do we really must to make them private?

Reply:
applied, getters were added